drivers: stepper: introduce stepper_ctrl.h for stepper motion control apis (#101001)
zephyrproject-rtos/zephyr: doc/releases/migration-guide-4.4.rst

* For :dtcompatible:`adi,tmc2209`, the property ``msx-gpios`` is now replaced by ``m0-gpios`` and
  ``m1-gpios`` for consistency with other step/dir stepper drivers.

* Since :github:`91979` & lateron :github:`101001`:

  * :c:func:`stepper_move_by` is replaced by :c:func:`stepper_ctrl_move_by`.
  * :c:func:`stepper_move_to` is replaced by :c:func:`stepper_ctrl_move_to`.
  * :c:func:`stepper_is_moving` is replaced by :c:func:`stepper_ctrl_is_moving`.
  * :c:func:`stepper_run` is replaced by :c:func:`stepper_ctrl_run`.
  * :c:func:`stepper_stop` is replaced by :c:func:`stepper_ctrl_stop`.
  * :c:func:`stepper_set_reference_position` is replaced by :c:func:`stepper_ctrl_set_reference_position`.
  * :c:func:`stepper_get_actual_position` is replaced by :c:func:`stepper_ctrl_get_actual_position`.

>   * :c:func:`stepper_set_microstep_interval` is replaced by :c:func:`stepper_ctrl_set_microstep_interval`.

Reviewer: ```suggestion
* Several API functions have been renamed:

  * ``stepper_move_by`` to :c:func:`stepper_ctrl_move_by`.
  * ``stepper_move_to`` to :c:func:`stepper_ctrl_move_to`.
  * ``stepper_is_moving`` to :c:func:`stepper_ctrl_is_moving`.
  * ``stepper_run`` to :c:func:`stepper_ctrl_run`.
  * ``stepper_stop`` to :c:func:`stepper_ctrl_stop`.
  * ``stepper_set_reference_position`` to :c:func:`stepper_ctrl_set_reference_position`.
  * ``stepper_get_actual_position`` to :c:func:`stepper_ctrl_get_actual_position`.
  * ``stepper_set_microstep_interval`` to :c:func:`stepper_ctrl_set_microstep_interval`.
```

> * The following events have been moved from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:

Author: The original stepper api was a monolithic api comprising of both motion controller and driver functions.
We have recently split that monolith into two seperate device driver classes, as a part of which some events have been moved from `stepper_event` enum class to `stepper_ctrl_event` enum class. The original `stepper_event` class still remains in the codebase, however only contains stepper hardware driver events now.

  * ``STEPPER_EVENT_STEPS_COMPLETED`` is replaced by ``STEPPER_CTRL_EVENT_STEPS_COMPLETED``.
  * ``STEPPER_EVENT_LEFT_END_STOP_DETECTED`` is replaced by ``STEPPER_CTRL_EVENT_LEFT_END_STOP_DETECTED```.
  * ``STEPPER_EVENT_RIGHT_END_STOP_DETECTED`` is replaced by ``STEPPER_CTRL_EVENT_RIGHT_END_STOP_DETECTED``.
  * ``STEPPER_EVENT_STOPPED`` is replaced by ``STEPPER_CTRL_EVENT_STOPPED``.

* ``step-gpios``, ``dir-gpios``, ``invert-direction`` and ``counter`` properties are removed
  from :dtcompatible:`adi,tmc2209`, :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`,

>   these are now are implemented by :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl`.

Reviewer: ```suggestion
* ``step-gpios``, ``dir-gpios``, ``invert-direction`` and ``counter`` properties have been removed
  from devicetree bindings :dtcompatible:`adi,tmc2209`, :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`,
  and moved to :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl` binding.
```

* :c:func:`stepper_ctrl_move_by`, :c:func:`stepper_ctrl_move_to`, :c:func:`stepper_ctrl_run`,
  :c:func:`stepper_ctrl_stop`, :c:func:`stepper_ctrl_is_moving`, :c:func:`stepper_ctrl_set_microstep_interval`
  and :c:func:`stepper_ctrl_set_event_cb` APIs are not implemented by :dtcompatible:`adi,tmc2209`,

>   :dtcompatible:`ti,drv84xx` and :dtcompatible:`allegro,a4979`.

Reviewer: What does this even mean? Were their software drivers implementing these API functions and now they are not? Are these functions optional?
The migration guide should only be talking about stuff affecting user apps (changes in public APIs, devicetree, and visible Kconfig options), and what the user has to do to get their app working as before.

* For :dtcompatible:`adi,tmc2209`, the property ``msx-gpios`` is now replaced by ``m0-gpios`` and
  ``m1-gpios`` for consistency with other step/dir stepper drivers.

* Since :github:`91979` & lateron :github:`101001`,

>   following APIs have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:

Reviewer 2: ```suggestion
  the following functions have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:
```

>   the following functions have been moved from :c:group:`stepper_driver_interface` to :c:group:`stepper_ctrl_interface`:

Author: done

  * :c:func:`stepper_move_by` is replaced by :c:func:`stepper_ctrl_move_by`.
  * :c:func:`stepper_move_to` is replaced by :c:func:`stepper_ctrl_move_to`.
  * :c:func:`stepper_is_moving` is replaced by :c:func:`stepper_ctrl_is_moving`.
  * :c:func:`stepper_run` is replaced by :c:func:`stepper_ctrl_run`.
  * :c:func:`stepper_stop` is replaced by :c:func:`stepper_ctrl_stop`.
  * :c:func:`stepper_set_reference_position` is replaced by :c:func:`stepper_ctrl_set_reference_position`.
  * :c:func:`stepper_get_actual_position` is replaced by :c:func:`stepper_ctrl_get_actual_position`.
  * :c:func:`stepper_set_microstep_interval` is replaced by :c:func:`stepper_ctrl_set_microstep_interval`.

> * Following events have been from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:

Reviewer 2: ```suggestion
* The following events have been moved from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:
```

  * ``stepper_set_microstep_interval`` to :c:func:`stepper_ctrl_set_microstep_interval`.

* The following events have been moved from :c:enum:`stepper_event` to :c:enum:`stepper_ctrl_event`:

  * ``STEPPER_EVENT_STEPS_COMPLETED`` to ``STEPPER_CTRL_EVENT_STEPS_COMPLETED``.
  * ``STEPPER_EVENT_LEFT_END_STOP_DETECTED`` to ``STEPPER_CTRL_EVENT_LEFT_END_STOP_DETECTED```.
  * ``STEPPER_EVENT_RIGHT_END_STOP_DETECTED`` to ``STEPPER_CTRL_EVENT_RIGHT_END_STOP_DETECTED``.
  * ``STEPPER_EVENT_STOPPED`` to ``STEPPER_CTRL_EVENT_STOPPED``.

* The ``step-gpios``, ``dir-gpios``, ``invert-direction``, and ``counter`` properties have been
  removed from step-dir stepper hardware driver bindings (:dtcompatible:`adi,tmc2209`, :dtcompatible:`ti,drv84xx`,
  and :dtcompatible:`allegro,a4979`) and moved to the new stepper motion controller binding
  :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl`.

* Motion control must now be performed through a :dtcompatible:`zephyr,gpio-step-dir-stepper-ctrl`
  device that references the stepper driver via the ``stepper-driver`` property. Applications must
  update their devicetree to add a motion controller node and use the ``stepper_ctrl_*`` APIs

>   instead of calling motion control functions directly on the driver device.

Reviewer: ```suggestion
  device that references the stepper hardware driver devicetree node via the ``stepper-driver`` property. Applications must
  update their devicetree to add a motion controller node and use the ``stepper_ctrl_*`` APIs,
  instead of calling motion control functions directly on the stepper hardware driver device.
```
